File: riveredge-frontend/src/apps/kuaiplm/services/prototype-build-sheet.ts
import { apiRequest } from '../../../services/api';

const BASE = '/apps/kuaiplm/prototype-build-sheets';

export type PrototypeBuildSheetStatus =
  | 'draft'
  | 'pending'
  | 'approved'
  | 'issued'
  | 'closed'
  | 'rejected';

export type PrototypeBuildRound = 'handboard' | 't1' | 't2' | 't3';

export interface PrototypeBuildAttachment {
  file_uuid: string;
  file_name?: string | null;
  media_kind?: 'text' | 'image' | 'file';
}

export interface PrototypeBuildSheet {
  id: number;
  uuid: string;
  sheet_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  round_key: PrototypeBuildRound | string;
  title: string;
  status: PrototypeBuildSheetStatus;
  project_requirements?: string | null;
  project_attachments?: PrototypeBuildAttachment[];
  electronics_requirements?: string | null;
  electronics_attachments?: PrototypeBuildAttachment[];
  electronics_status: string;
  structure_requirements?: string | null;
  structure_attachments?: PrototypeBuildAttachment[];
  structure_status: string;
  manufacturing_opinion?: string | null;
  quality_opinion?: string | null;
  remarks?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  issued_at?: string | null;
  closed_at?: string | null;
  created_at: string;
  updated_at: string;
  created_by?: number | null;
  created_by_name?: string | null;
  updated_by?: number | null;
  updated_by_name?: string | null;
}

export const prototypeBuildSheetApi = {
  list(params?: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    project_id?: number;
    round_key?: string;
  }) {
    return apiRequest<{ items: PrototypeBuildSheet[]; total: number }>(BASE, { params });
  },
  get(id: number) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}`);
  },
  create(data: {
    project_id: number;
    title: string;
    sheet_code?: string;
    round_key?: string;
    project_requirements?: string;
    project_attachments?: PrototypeBuildAttachment[];
    remarks?: string;
  }) {
    return apiRequest<PrototypeBuildSheet>(BASE, { method: 'POST', body: data });
  },
  update(
    id: number,
    data: {
      title?: string;
      project_requirements?: string;
      project_attachments?: PrototypeBuildAttachment[];
      remarks?: string;
    },
  ) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}`, { method: 'PUT', body: data });
  },
  updateSection(
    id: number,
    section: 'electronics' | 'structure',
    data: { requirements?: string; attachments?: PrototypeBuildAttachment[] },
  ) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/sections/${section}`, {
      method: 'PUT',
      body: data,
    });
  },
  updateSignoff(
    id: number,
    data: { manufacturing_opinion?: string; quality_opinion?: string },
  ) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/signoff`, {
      method: 'PUT',
      body: data,
    });
  },
  submit(id: number) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/submit`, { method: 'POST' });
  },
  approve(id: number) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/approve`, { method: 'POST' });
  },
  reject(id: number) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/reject`, { method: 'POST' });
  },
  issue(id: number) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/issue`, { method: 'POST' });
  },
  close(id: number) {
    return apiRequest<PrototypeBuildSheet>(`${BASE}/${id}/close`, { method: 'POST' });
  },
  remove(id: number) {
    return apiRequest<void>(`${BASE}/${id}`, { method: 'DELETE' });
  },
};

File: riveredge-backend/src/apps/kuaiplm/schemas/prototype_build_sheet.py
"""样机制作书 Schema（研发项目 §2.15）"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PrototypeBuildAttachment(BaseModel):
    file_uuid: str = Field(..., min_length=1, max_length=36)
    file_name: Optional[str] = Field(None, max_length=200)
    media_kind: str = Field(default="file", description="text/image/file")


class PrototypeBuildSheetCreate(BaseModel):
    project_id: int
    title: str = Field(..., min_length=1, max_length=200)
    sheet_code: Optional[str] = Field(None, max_length=50)
    round_key: str = Field(default="t1", max_length=20)
    project_requirements: Optional[str] = None
    project_attachments: List[PrototypeBuildAttachment] = Field(default_factory=list)
    remarks: Optional[str] = None


class PrototypeBuildSheetUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    project_requirements: Optional[str] = None
    project_attachments: Optional[List[PrototypeBuildAttachment]] = None
    remarks: Optional[str] = None


class PrototypeBuildSectionUpdate(BaseModel):
    requirements: Optional[str] = None
    attachments: List[PrototypeBuildAttachment] = Field(default_factory=list)


class PrototypeBuildSignoffUpdate(BaseModel):
    manufacturing_opinion: Optional[str] = None
    quality_opinion: Optional[str] = None


class PrototypeBuildSheetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    sheet_code: str
    project_id: int
    project_code: str
    project_name: str
    round_key: str
    title: str
    status: str
    project_requirements: Optional[str] = None
    project_attachments: List[dict] = Field(default_factory=list)
    electronics_requirements: Optional[str] = None
    electronics_attachments: List[dict] = Field(default_factory=list)
    electronics_status: str
    structure_requirements: Optional[str] = None
    structure_attachments: List[dict] = Field(default_factory=list)
    structure_status: str
    manufacturing_opinion: Optional[str] = None
    quality_opinion: Optional[str] = None
    remarks: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    issued_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None


class PrototypeBuildSheetListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    sheet_code: str
    project_id: int
    project_code: str
    project_name: str
    round_key: str
    title: str
    status: str
    electronics_status: str
    structure_status: str
    created_at: datetime
    updated_at: datetime
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class PrototypeBuildSheetListResponse(BaseModel):
    items: List[PrototypeBuildSheetListItem]
    total: int
